#Basic Data Summary:

We have been provided with data to see churn rates of a certain telecom company. Creating this model will give us insights on which factors affect a company’s churn and how can we improve the current churn rates, which will provide more profit to the company.


# Importing Files

In [0]:
from google.colab import files
files.upload()

Saving GA_Dataset.xlsx to GA_Dataset.xlsx


{'GA_Dataset.xlsx': b'PK\x03\x04\x14\x00\x06\x00\x08\x00\x00\x00!\x00\x8c\xe9\x8c\x8c~\x01\x00\x00\x8c\x05\x00\x00\x13\x00\xdc\x01[Content_Types].xml \xa2\xd8\x01(\xa0\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\

In [0]:
!ls

GA_Dataset.xlsx  sample_data


# Reading Data

In [0]:
import pandas as pd

xlsx = pd.ExcelFile("GA_Dataset.xlsx")
data = pd.read_excel(xlsx, "Data")
dataM = data

data

Churn  AccountWeeks  ContractRenewal  ...  MonthlyCharge  OverageFee  RoamMins
0         0           128                1  ...           89.0        9.87      10.0
1         0           107                1  ...           82.0        9.78      13.7
2         0           137                1  ...           52.0        6.06      12.2
3         0            84                0  ...           57.0        3.10       6.6
4         0            75                0  ...           41.0        7.42      10.1
...     ...           ...              ...  ...            ...         ...       ...
3328      0           192                1  ...           71.7       10.78       9.9
3329      0            68                1  ...           56.4        7.67       9.6
3330      0            28                1  ...           56.0       14.44      14.1
3331      0           184                0  ...           50.0        7.98       5.0
3332      0            74                1  ...          100.0       13.30      13.7

[3333 rows x 11 columns]

# Data Cleaning and Manipulation

Firstly, we need to see if there are any null values and if we need to drop or impute the data, depending on the quantity and the type of data

In [0]:
data.dropna()
data

Churn  AccountWeeks  ContractRenewal  ...  MonthlyCharge  OverageFee  RoamMins
0         0           128                1  ...           89.0        9.87      10.0
1         0           107                1  ...           82.0        9.78      13.7
2         0           137                1  ...           52.0        6.06      12.2
3         0            84                0  ...           57.0        3.10       6.6
4         0            75                0  ...           41.0        7.42      10.1
...     ...           ...              ...  ...            ...         ...       ...
3328      0           192                1  ...           71.7       10.78       9.9
3329      0            68                1  ...           56.4        7.67       9.6
3330      0            28                1  ...           56.0       14.44      14.1
3331      0           184                0  ...           50.0        7.98       5.0
3332      0            74                1  ...          100.0       13.30      13.7

[3333 rows x 11 columns]

This shows us that there are no null values present in the data. So we go ahead with first looking at a summary of all attributes that have been provided to us.

In [0]:
data.describe()

Churn  AccountWeeks  ...   OverageFee     RoamMins
count  3333.000000   3333.000000  ...  3333.000000  3333.000000
mean      0.144914    101.064806  ...    10.051488    10.237294
std       0.352067     39.822106  ...     2.535712     2.791840
min       0.000000      1.000000  ...     0.000000     0.000000
25%       0.000000     74.000000  ...     8.330000     8.500000
50%       0.000000    101.000000  ...    10.070000    10.300000
75%       0.000000    127.000000  ...    11.770000    12.100000
max       1.000000    243.000000  ...    18.190000    20.000000

[8 rows x 11 columns]

Now that we have got the Mean, interquartile range, etc. We can move ahead to plot some graphs and check and analyse the data

# Exploratory Data Analysis

Since we could see that there was not any cleaning required (Data had no missing or null values and it is labelled properly for training the model). We go ahead and try to get some insights first before running any models.

But first, let’s import some libraries to plot those graphs:


In [0]:
import plotly.express as px

Now that out of the way, lets first analyse the individual attributes we have.
We have attributes namely, 
- Churn (shows whether the customer has left the company or not), 
- AccountWeeks (shows the number of weeks customer has had an active account)
- ContractRenewal (in layman terms, whether the customer has done any recharge to continue using the services)
- DataPlan (shows whether the customer has an active Data Plan or not)
- DataUsage (shows monthly usage in GBs)
- CustServCalls (shows the number of calls that went to customer service)
- DaysMins (shows average time a customer spends on his/her calls)
- DayCalls (shows the number of times customer makes a call in a day)
- Monthly Charge (gives us their average monthly bill)
- Overage Fee (shows us the largest Overage Fee that the customer has incurred in the last 12 months)
- RoamMins (gives us the average number of Roam Mins)

**Univariate Analysis:**


So let's start with Account Weeks, which shows the number of weeks customer has had an active account.

This can be a good parameter to look into as it shows us Loyalty of a customer towards the brand. The more the time a customer has used the brand for, there are really fewer chances that they'll switch to any other brand.


In [0]:
px.histogram(data, x = "AccountWeeks")

This Histogram Plot shows us that around 100 customers (see 50-150) have stayed longer than 500 weeks in that company.

Now let's come over to the Data bit. First Let's see how many customers have a data plan in the first place.

In [0]:
px.histogram(data, x="DataPlan")

This shows us that the majority of the customer base does not have an active data plan.

So, with the non-data customers out of the way, let's see how much data is consumed by a person with an active data plan.

In [0]:
data = data[data["DataPlan"] != 0]

px.histogram(data, x="DataUsage")

We can see that the majority of customer base use around 3gb of data and the data plans around it. The company can introduce more plans like this to introduce more loyal customers. 

**Bivariate Analysis:**

Now let's try and put two attributes together and see how they're related to each other.

For starters let's try and see how many people who use data, call customer care executives in a month

In [0]:
px.scatter(x=data["CustServCalls"], y=data["DataUsage"])

This shows us that people who do not use data or use less than 1gb per day call customer care more frequently than the ones who use more data.

Now the main purpose of plotting this is to see whether people who call customer care more often are still customers or not

In [0]:
px.scatter(x = data["Churn"], y = data["CustServCalls"])

This shows us that people who call customer care more often have left the company.

Now let's compare for how much time people talk in a day vs the people who use data

In [0]:
px.scatter(x=data["DataUsage"], y=data["DayMins"])

Now let's try and see if people who spend a lot of time talking over the phone, leave the carrier or not

In [0]:
px.scatter(x = data["Churn"], y = data["DayMins"])

This shows us that people who spend more time on the phone are less likely to leave the company and switch to another carrier.

**Correlation Graph**

Finally, let’s see what attributes are correlated in our data and which one we should select in our model.

In [0]:
import plotly.figure_factory as ff
corrs = dataM.corr()
figure = ff.create_annotated_heatmap(
    z=corrs.values,
    x=list(corrs.columns),
    y=list(corrs.index),
    annotation_text=corrs.round(2).values,
    showscale=True)

figure

If we compare "Churn" with other attributes, we see that "DataPlan" and "DataUsage" are somehow correlated to "Churn".

Also, now-a-days there are no roaming charges or any charges related to roaming being applied to any customer on any tele company. So, we remove that from the equation as well.

Finally, coming to the last attribute, "CustServCalls" gave us some mixed insights as well as inaccurate results. Hence, removing that attribute as well.

In [0]:
data.pop("RoamMins")
data.pop("DataPlan")
data.pop("DataUsage")
data.pop("CustServCalls")
data.columns

Index(['Churn', 'AccountWeeks', 'ContractRenewal', 'DayMins', 'DayCalls',
       'MonthlyCharge', 'OverageFee'],
      dtype='object')

# Training the model

In [0]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn import preprocessing

y = data["Churn"]
x = data
x.pop("Churn")

norm_x = preprocessing.normalize(x)

trainx, testx, trainy, testy = train_test_split(norm_x, y, test_size = 0.3)      

# Logistic Regression

In [0]:
import numpy as np
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn import metrics

logreg = LogisticRegression()
logreg.fit(trainx, trainy)

y_pred = logreg.predict(testx)
print(np.sqrt(mean_squared_error(testy,y_pred))) 
score = logreg.score(testx, testy)
print(score)

cm = metrics.confusion_matrix(testy, y_pred)
print(cm)

0.3122065409111821
0.9025270758122743
[[250   0]
 [ 27   0]]


The RMSE value is 0.31. This means that the average difference from the best fit line for the Churn is 0.31.

And our accuracy in this model is 90.25%